File: leetcode-04/solutions/python3/1019.py
class Solution:
    def nextLargerNodes(self, head: ListNode) -> List[int]:
        ans = []
        stack = []
        curr = head

        while curr:
            while stack and ans[stack[-1]] < curr.val:
                ans[stack[-1]] = curr.val
                stack.pop()
            stack.append(len(ans))
            ans.append(curr.val)
            curr = curr.next

        for i in stack:
            ans[i] = 0

        return ans

File: leetcode-04/solutions/cpp/1019.cpp
class Solution {
 public:
  vector<int> nextLargerNodes(ListNode* head) {
    vector<int> ans;
    vector<int> stack;

    for (auto curr = head; curr; curr = curr->next) {
      while (stack.size() && ans[stack.back()] < curr->val) {
        ans[stack.back()] = curr->val;
        stack.pop_back();
      }
      stack.push_back(ans.size());
      ans.push_back(curr->val);
    }

    for (int i : stack) ans[i] = 0;

    return ans;
  }
};